### PR TITLE
feat: Integrated channels content metadata transmissions V2

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
details written below, changes to come once enterprise catalog PR goes live.

All questions/concerns welcome

Old implementation:
old IC

new proposed implementation:
new IC

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+**querying all content from the catalog service is expensive**- Asking for all content under the catalog, for every
+customer, for every catalog, on every transmission is not only an undesirably large request to be sending, but also
+consumes much of the enterprise catalog's DB load. We need a smarter solution to how the Integrated Channels requests
+content from the catalog service.
+
+**putting the responsibility of determining differences** between what's already been sent to the customers and what's
+currently in customer's catalogs adds much complexity to the integrated channels which can make it unapproachable from
+an onboarding perspective but also from a debugging point of view. Log noise and number of potential error sources means
+tracking down issues within the Integrated Channels are costly and time consuming.
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:32]: Nit: "are costly and time consuming" -> "are time consuming and thus, costly."  (time is money).

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+How do the Integrated Channels currently handle customer catalog data?
+======================================================================
+
+existing flow chart:
+
+.. image:: ../images/0009-moving-catalog-diffing-to-enterprise-catalog.rst
+  :width: 500
+
+Currently on a scheduled job, the Integrated Channels' content metadata exporter iterates over each customer's catalog
+and builds up a dict of content representing the current state of the customer's catalog content. Take away- as it
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:43]: "Take away-" -> "Take away:"

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+How the Integrated Channels could better determine if customer metadata needs updating
+======================================================================================
+
+proposed flow chart:
+
+.. image:: ../images/new-integrated-channels-metadata-flow.png
+  :width: 500
+
+If we build an endpoint that takes catalog UUIDs and a set of content keys linked to the uuid and returns three buckets
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:58]: UUID or UUIDs?

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+of data:
+
+1) What content keys exist under the specified catalog that were not provided in the list of content keys
+
+2) What content keys don't exist under the specified catalog that were provided in the list of content keys
+
+3) What the last updated at times of the content keys are for the content keys provided that are under the specified
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:65]: This is the primary use case for the endpoint so I think we should list it first versus third.

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+catalog
+
+We would be able to move the responsibility of diffing what the customer has with what the customer should have to the
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:68]: It's not technically diffing what the "customer has" although that's one use case for the endpoint.

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+times of the courses that already exist to get which courses need updates. Creates and deletes would already determined
+by the first endpoint.
+
+Further Improvements
+====================
+
+There are two additional improvements to explore with the Integrated Channels' metadata transmission flow. Firstly,
+we remove records from our transmission audit table when a delete request is issued. This can result in `lost` nodes of
+content (as we've actively seen from customers) where content can be assumed to be deleted on our end but have it exist
+still on the customer's external LMS. The fix here is pretty easy- if we choose to not delete records, but rather create
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:79]: We really should consider doing this asap -- imho this is "need to have" versus "nice to have".

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+a new, nullable `deleted_at` or equivalent field in the transmission record table. Then we would be able to mark and
+exclude any deleted records from appropriate look ups, but still have something to help us identify past courses that
+were sent to customers.
+
+As stated earlier, currently the only method to retrieve content under a customer's catalog is the singular, bulk
+`get_content_metadata` endpoint. This returns all content metadata belonging to an enterprise catalog. If we were to
+build out the existing endpoint where individual content key's worth of metadata could be specified in the request body,
```
Comment (mattdrayer) [inline docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:86]: One more improvement to list -- being able to take an "after X" date value as a query parameter and return only those records which have been created/updated/deleted after that date.  We could consider a "before X" date value as well, which would enable additional scenarios such as a "between" type lookup.

**Review by mattdrayer — APPROVED** [review]

File: enterprise/api_client/enterprise_catalog.py [hunk 8]
```diff
@@ -107,6 +135,14 @@ def get_content_metadata_url(uuid):
         return EnterpriseCatalogApiClient.API_BASE_URL + \
             EnterpriseCatalogApiClient.GET_CONTENT_METADATA_ENDPOINT.format(uuid)
 
+    @staticmethod
```
Comment (alex-sheehan-edx) [inline enterprise/api_client/enterprise_catalog.py:138]: unsure if this is needed or just a good to have for the future  ¯\_(ツ)_/¯

File: integrated_channels/sap_success_factors/exporters/content_metadata.py [hunk 9]
```diff
@@ -253,11 +222,42 @@ def transform_program_key(self, content_metadata_item):
         """
         return content_metadata_item['uuid']
 
-    def get_content_id(self, content_metadata_item):
+    def _get_content_id(self, content_metadata_item):
```
Comment (alex-sheehan-edx) [inline integrated_channels/sap_success_factors/exporters/content_metadata.py:225]: standardizing helper methods in the file

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+1) What the last updated at times of the content keys are for the content keys provided that are under the specified
+catalog
+
+2) What content keys exist under the specified catalog that were not provided in the list of content keys
+
+3) What content keys don't exist under the specified catalog that were provided in the list of content keys
+
+We would be able to move the responsibility of diffing what we believe the customer has with what the customer should
+have, to the enterprise catalog. This would mean that the only job of the Integrated Channels would be to compare the
+last updated times of the courses that already exist to get which courses need updates. Creates and deletes would
```
[the remark this answers is not in the document]
Comment (binodpant) [reply docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:70]: or if it's not in scope for this PR that's cool too..

[the remark this answers is not in the document]
Comment (binodpant) [reply docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:70]: hmm okay sorry still did not get clear on: we will still fetch all content metadata, post the enterprise-catalog PR, like we used to do pre-optimization, or will there be a change in what metadata items we get? Either answer is fine, just wanted to be clear on which one we do for now, yeah sounds like we may need a future ticket to only get what we want .. kind of work

**Comment by binodpant** [comment]
Sounds like not all of the ADR is being included in this and the upcoming PR, but a subset of it, it will be good to indicate in the PR desc which ones those are mainly for later revisiting / log trail

File: enterprise/api_client/enterprise_catalog.py [hunk 11]
```diff
@@ -99,6 +100,60 @@ def get_enterprise_catalog(self, catalog_uuid, should_raise_exception=True):
                 )
             return {}
 
+    @JwtLmsApiClient.refresh_token
+    def get_catalog_diff(self, enterprise_customer_catalog, content_keys, should_raise_exception=True):
+        """
+        Gets the representational difference between a list of course keys and the current state of content under an
+        enterprise catalog. This difference it returned as three buckets of data: `items_not_found`,
```
Comment (binodpant) [inline enterprise/api_client/enterprise_catalog.py:107]: I think one of these needs to be `items_found` ?

File: enterprise/api_client/enterprise_catalog.py [hunk 12]
```diff
@@ -99,6 +100,60 @@ def get_enterprise_catalog(self, catalog_uuid, should_raise_exception=True):
...
+        items_to_create = []
+        matched_items = []
+        try:
+            response = endpoint.post(body)
+            items_to_delete = response.get('items_not_found')
+            items_to_create = response.get('items_not_included')
+            matched_items = response.get('items_found')
+
+        except (SlumberBaseException, ConnectionError, Timeout) as exc:
+            if should_raise_exception:
```
[the remark this answers is not in the document]
Comment (binodpant) [reply enterprise/api_client/enterprise_catalog.py:149]: alternatively if the intent is to decide whether to raise exception or not, I am not sure why we need that either

File: enterprise/api_client/enterprise_catalog.py [hunk 13]
```diff
@@ -107,6 +162,14 @@ def get_content_metadata_url(uuid):
         return EnterpriseCatalogApiClient.API_BASE_URL + \
             EnterpriseCatalogApiClient.GET_CONTENT_METADATA_ENDPOINT.format(uuid)
 
+    @staticmethod
+    def get_catalog_diff_url(uuid):
+        """
+        Get the url for generating a diff between content keys and enterprise catalog metadata
+        """
+        return EnterpriseCatalogApiClient.API_BASE_URL + \
```
Comment (binodpant) [inline enterprise/api_client/enterprise_catalog.py:170]: nit: f-strings vs format.

File: integrated_channels/integrated_channel/models.py [hunk 14]
```diff
@@ -308,6 +308,11 @@ class ContentMetadataItemTransmission(TimeStampedModel):
         blank=True,
         null=True,
     )
+    deleted_at = models.DateTimeField(
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/integrated_channel/models.py:311]: but if so this means our logic on how we ind things must change, guessing you already have it in the queryset somewhere... meaning exclude deleted ones .. or sth like that

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:70]: we definitely need a follow up ticket. This work is closer to a more accurate pre-optimization than what we were doing before. It also let's us now easily implement a more efficient, filtered content_metadata endpoint in the future.

File: enterprise/api_client/enterprise_catalog.py [hunk 15]
```diff
@@ -99,6 +100,60 @@ def get_enterprise_catalog(self, catalog_uuid, should_raise_exception=True):
...
+        Gets the representational difference between a list of course keys and the current state of content under an
+        enterprise catalog. This difference it returned as three buckets of data: `items_not_found`,
+        `items_not_included` and `items_not_found`.
+
+        Arguments:
+            enterprise_customer_catalog (EnterpriseCustomerCatalog): The catalog object who's content is being diffed.
+            content_keys (list): List of string content keys
+            should_raise_exception (Bool): Optional param for whether or not api response exceptions should be raised.
+
+        Response:
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/api_client/enterprise_catalog.py:115]: I'm going to change this to the response of the method, so in this case no it's three separate values

File: tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_content_metadata.py [hunk 16]
```diff
@@ -103,7 +104,7 @@ def test_transmit_create_failure(self, update_content_metadata_mock):
 
     @responses.activate
     @mock.patch('integrated_channels.sap_success_factors.client.SAPSuccessFactorsAPIClient.update_content_metadata')
-    def test_transmit_api_usage_limit(self, update_content_metadata_mock):
+    def test_transmit_api_usage_limit_x(self, update_content_metadata_mock):
```
Comment (binodpant) [inline tests/test_integrated_channels/test_sap_success_factors/test_transmitters/test_content_metadata.py:107]: is the _x a hint to pylint, or an actual test intended name?

File: integrated_channels/sap_success_factors/transmitters/content_metadata.py [hunk 17]
```diff
@@ -62,17 +67,17 @@ def transmit(self, payload, **kwargs):
...
         # Since, chunk_items is a generator and we have already iterated through the items that need to
         # be transmitted. Rest of the items are the ones that need to marked as not transmitted.
         for chunk in chunk_items:
             chunked_items = list(chunk.values())
-            self._remove_failed_items(chunked_items, items_to_create, items_to_update, items_to_delete)
+            self._remove_failed_items(chunked_items, create_payload, items_to_update, items_to_delete)
 
-        self._create_transmissions(items_to_create, items_catalog_updated_times, content_catalog_map)
-        self._update_transmissions(items_to_update, transmission_map, items_catalog_updated_times)
+        self._create_transmissions(create_payload, content_updated_mapping)
```
Comment (binodpant) [inline integrated_channels/sap_success_factors/transmitters/content_metadata.py:79]: I think having the ordering be consistent between the methods would be great, for ex the content_update_mapping being last or sth like that

File: enterprise/api_client/enterprise_catalog.py [hunk 18]
```diff
@@ -99,6 +100,61 @@ def get_enterprise_catalog(self, catalog_uuid, should_raise_exception=True):
                 )
             return {}
 
+    @JwtLmsApiClient.refresh_token
+    def get_catalog_diff(self, enterprise_customer_catalog, content_keys, should_raise_exception=True):
+        """
+        Gets the representational difference between a list of course keys and the current state of content under an
+        enterprise catalog. This difference is returned as three buckets of data: `items_not_found`,
+        `items_not_included` and `items_found`.
```
Comment (adamstankiewicz) [inline enterprise/api_client/enterprise_catalog.py:108]: nit: is `items_found` referring to the `matched_items` used below? Should we converge on one or the other?

File: integrated_channels/integrated_channel/exporters/content_metadata.py [hunk 19]
```diff
@@ -75,97 +70,189 @@ def __init__(self, user, enterprise_configuration):
...
+        Retrieve all past content metadata transmission records that have a `content_id` contained within a provided
+        list.
+        """
+        ContentMetadataItemTransmission = apps.get_model(
+            'integrated_channel',
+            'ContentMetadataItemTransmission'
+        )
+        content_keys = [item.get('content_key') for item in items]
+
+        # Do we want to filter out deleted items here?
```
Comment (adamstankiewicz) [inline integrated_channels/integrated_channel/exporters/content_metadata.py:169]: Is this still an open question?

File: docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst [hunk 20]
```diff
@@ -0,0 +1,104 @@
+Moving the responsibility of catalog diffing to the enterprise catalog
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply docs/decisions/0009-moving-catalog-diffing-to-enterprise-catalog.rst:1]: oh this is weird... I thought I updated this doc but the preview isn't changing

File: integrated_channels/integrated_channel/exporters/content_metadata.py (hunk 19 above)
Comment (alex-sheehan-edx) [reply integrated_channels/integrated_channel/exporters/content_metadata.py:169]: It was when writing, I think I am more confident in the state now.

**Pull request merged**